// cli-service/pack.config.template.js
var path = require('path')
module.exports = {
  webpackConfig: {
    context: __dirname,
    entry: {
      app: path.resolve(__dirname, 'src/app/index.ts')
    },
    resolveLoader: {
      modules: [path.resolve(__dirname, 'src'), path.resolve(__dirname, 'node_modules')]
    },
    resolve: {
      modules: [path.resolve(__dirname, 'src'), path.resolve(__dirname, 'node_modules')],
      alias: {
        app: path.resolve('src/app')
      }
    }
  },
  common: {
    title: 'vue template',
    indexTemplate: path.resolve(__dirname, 'src/index.html'),
    indexOutput: path.resolve(__dirname, 'dist/index.html'),
    staticOriginDirctory: path.resolve(__dirname, 'src/app/static'),
    assetsRoot: path.resolve(__dirname, 'dist'),
    assetsSubDirectory: 'static',
    assetsPublicPath: '//acdnpy.pandateacher.com/',
    // Run the build command with an extra argument to
    // View the bundle analyzer report after build finishes:
    // `npm run build --report`
    // Set to `true` or `false` to always turn it on or off
    bundleAnalyzerReport: false
  },
  production: {
    sourceMap: true,
    bundleAnalyzerReport: false
  },
  debug: {
    sourceMap: true,
    bundleAnalyzerReport: false
  },
  development: {
    port: 8085,
    indexOutput: 'index.html',
    autoOpenBrowser: false,
    assetsSubDirectory: 'static',
    assetsPublicPath: '/',
    proxyTable: {},
    // CSS Sourcemaps off by default because relative paths are "buggy"
    // with this option, according to the CSS-Loader README
    // (https://github.com/webpack/css-loader#sourcemaps)
    // In our experience, they generally work as expected,
    // just be aware of this issue when enabling this option.
    sourceMap: true
  }
}
