// cli-service/lib/config/.postcssrc.js
// https://github.com/michael-ciniawsky/postcss-load-config

module.exports = {
  parser: 'postcss-scss',
  plugins: [
    // to edit target browsers: use "browserslist" field in package.json
    require('postcss-functions')({
      functions: {
        pr(value) {
          return (value.replace('px', '') / 75) * 1 + 'rem'
        },
        rgba(color, opacity) {
          if (arguments.length == 4) {
            return `rgba(${arguments[0]}, ${arguments[1]}, ${arguments[2]}, ${
              arguments[3]
            })`
          } else {
            color = color
              .split(':')[1]
              .trim()
              .slice(1)
            opacity = opacity.split(':')[1].trim()
            if (color.length == 3) {
              var temp = color.split('')
              color = temp.concat(temp).join('')
            }
            const block = color.length / 3
            const r = parseInt(color.slice(0, block), 16)
            const g = parseInt(color.slice(block, block * 2), 16)
            const b = parseInt(color.slice(block * 2), 16)
            return `rgba(${r}, ${g}, ${b}, ${opacity})`
          }
        }
      }
    }),
    require('postcss-mixins')({
      mixins: {
        fz(mixin, value) {
          return {
            '&': {
              'font-size': `${value}px`
            },
            '[data-dpr="2"] &': {
              'font-size': `${value * 2}px`
            },
            '[data-dpr="3"] &': {
              'font-size': `${value * 3}px`
            }
          }
        }
      }
    }),
    require('precss'),
    require('postcss-cssnext')({
      features: {
        rem: false
      }
    })
  ]
}
