// cli-service/lib/config/common.js
const path = require('path')
const HappyPack = require('happypack')
const ProgressPlugin = require('progress-bar-webpack-plugin')
const utils = require('./utils')
const vueLoaderConfig = require('./vueLoader')
const getConfig = require('../util/getConfig')
const resolveProject = require('../util/resolveProject')

module.exports = {
  resolveLoader: { // loader 查找路径
    modules: [path.resolve(__dirname, '../../node_modules')]
  },
  output: {
    path: getConfig('assetsRoot'),
    filename: '[name].js',
    publicPath: getConfig('assetsPublicPath')
  },
  resolve: {
    modules: [path.resolve(__dirname, '../../node_modules')], // webpack 中的模块查询范围
    extensions: ['.js', '.ts', '.vue', '.json']
  },
  module: {
    rules: [
      {
        test: /\.js$/,
        loader: 'happypack/loader?id=js&verbose=false',
        include: [resolveProject('src')]
      },
      {
        test: /\.vue$/,
        loader: 'vue-loader',
        options: vueLoaderConfig
      },
      {
        test: /\.(png|jpe?g|gif|svg)(\?.*)?$/,
        loader: 'url-loader',
        options: {
          limit: 10000,
          name: utils.assetsPath('img/[name].[hash:7].[ext]')
        }
      },
      {
        test: /\.(mp4|webm|ogg|mp3|wav|flac|aac)(\?.*)?$/,
        loader: 'url-loader',
        options: {
          limit: 10000,
          name: utils.assetsPath('media/[name].[hash:7].[ext]')
        }
      },
      {
        test: /\.(woff2?|eot|ttf|otf)(\?.*)?$/,
        loader: 'url-loader',
        options: {
          limit: 10000,
          name: utils.assetsPath('fonts/[name].[hash:7].[ext]')
        }
      }
    ]
  },
  plugins: [
    new HappyPack({
      id: 'js',
      loaders: ['babel-loader']
    }),
    new HappyPack({
      id: 'ts',
      // loaders is the only required parameter:
      loaders: [
        {
          loader: 'ts-loader',
          options: {
            appendTsSuffixTo: [/\.vue$/],
            happyPackMode: true
          }
        }
      ]
    }),
    new ProgressPlugin()
  ]
}
